### PR TITLE
irmin-bench: CLI and PP for stat trace

### PR DESCRIPTION
**Pull request by Ngoguey42**
Last PR extracted from #1332

I'm not totally happy with `trace_stat_summary_pp.ml`. I will improve it when new features are needed.

A single summary pretty print:
```
dune exec -- ./bench/irmin-pack/trace_stats.exe pp /data/ngoguey/irmin/_artefacts/master_none/boostrap_summary.json
Hostname        | comanche
Word Size       | 64 bits
Start Time      | 2021/04/15 19:12:54 (GMT)
Inode Config    | mls:32 bf:32 sh:256
Store Type      | pack
Path Conversion | none

Block played count *C                  | 0 (before) |   335874   |   671748   |  1007622   | 1343496 (end)
Blocks progress *C                     |     0%     |    25%     |    50%     |    75%     |     100%
                                       |            |            |            |            |
Wall time elapsed *C                   |        0us |    3h53min |   13h34min |       1d7h |         2d14h
CPU time elapsed *C                    |        0us |    3h50min |   13h22min |       1d6h |         2d13h
                                       |            |            |            |            |
Ops count *C                           |    0.000 M |  178.693 M |  404.025 M |  693.862 M |    1089.854 M
Read only ops per sec *LA              |        n/a |   6451.800 |   3888.361 |   3024.875 |      2492.405
Read write ops per sec *LA             |        n/a |   2432.414 |   1408.054 |   1024.541 |       784.451
                                       |            |            |            |            |
Commit every sec *LA                   |        n/a |     14.895 |      7.110 |      4.294 |         2.344
Add every sec *LA                      |        n/a |   2417.993 |   1383.707 |   1014.295 |       773.094
Remove every sec *LA                   |        n/a |     14.363 |     24.112 |     10.230 |        11.123
Find every sec *LA                     |        n/a |   4845.606 |   2700.683 |   2126.733 |      1663.274
Mem every sec *LA                      |        n/a |   1114.650 |    906.733 |    756.435 |       751.784
Mem_tree every sec *LA                 |        n/a |    491.544 |    280.945 |    141.708 |        77.346
Copy every sec *LA                     |        n/a |  0.057_881 |  0.234_617 |  0.016_686 |     0.233_792
                                       |            |            |            |            |
Add count per block *LA                |        n/a |    162.336 |    194.616 |    236.202 |       329.843
Remove count per block *LA             |        n/a |  0.964_275 |  3.391_339 |  2.382_319 |     4.745_588
Find count per block *LA               |        n/a |    325.317 |    379.847 |    495.259 |       709.640
Mem count per block *LA                |        n/a |     74.834 |    127.531 |    176.153 |       320.751
Mem_tree count per block *LA           |        n/a |     33.001 |     39.514 |     33.000 |        33.000
Copy count per block *LA               |        n/a |  0.003_886 |  0.032_999 |  0.003_886 |     0.099_748
                                       |            |            |            |            |
Block duration *LA                     |   0.000  s |   0.067  s |   0.141  s |   0.233  s |      0.427  s
Buildup seen duration *LA              |   0.000 ms |   5.401 ms |   6.814 ms |   8.866 ms |     12.523 ms
Buildup unseen duration *LA            |        n/a |   1.893 ms |   2.330 ms |   2.906 ms |      4.478 ms
Commit duration *LA                    |        n/a |   0.060  s |   0.132  s |   0.221  s |      0.410  s
                                       |            |            |            |            |
Add duration *LA                       |        n/a |  11.945 µs |  11.188 µs |   9.969 µs |      9.556 µs
Remove duration *LA                    |        n/a |  11.436 µs |  12.412 µs |  34.320 µs |     32.297 µs
Find duration *LA                      |        n/a |   9.389 µs |   9.486 µs |   9.294 µs |      9.272 µs
Mem duration *LA                       |        n/a |   4.159 µs |   6.907 µs |   9.856 µs |      7.912 µs
Mem_tree duration *LA                  |        n/a |   2.439 µs |   2.659 µs |   2.597 µs |      2.829 µs
Copy duration *LA                      |        n/a |  15.937 µs |  27.897 µs |  34.672 µs |     13.476 µs
Checkout duration *LA                  |        n/a |   4.294 µs |   4.816 µs |   4.282 µs |      5.624 µs
                                       |            |            |            |            |
pack.finds per block *LA               |        n/a |   1610.005 |   2046.639 |   2303.584 |      2859.420
pack.cache_misses per block *LA        |        n/a |     79.175 |    132.383 |    134.764 |       206.057
pack.appended_hashes per block *LA     |        n/a |  0.035_388 |  0.093_458 |  0.165_040 |     0.073_375
pack.appended_offsets per block *LA    |        n/a |   1628.925 |   2160.399 |   2614.423 |      3576.880
                                       |            |            |            |            |
tree.contents_hash per block *LA       |        n/a |    125.817 |    141.760 |    150.103 |       176.406
tree.contents_find per block *LA       |        n/a |    185.714 |    217.273 |    263.228 |       348.266
tree.contents_add per block *LA        |        n/a |     69.662 |     77.439 |     88.737 |       107.554
tree.node_hash per block *LA           |        n/a |    210.284 |    269.217 |    310.486 |       412.139
tree.node_mem per block *LA            |        n/a |    209.476 |    266.932 |    308.375 |       407.885
tree.node_add per block *LA            |        n/a |    206.451 |    263.796 |    305.862 |       404.644
tree.node_find per block *LA           |        n/a |    679.684 |    798.250 |    819.762 |       965.929
tree.node_val_v per block *LA          |        n/a |  7.873e-11 |  7.508e-17 |  7.160e-23 |     6.829e-29
tree.node_val_find per block *LA       |        n/a |          0 |          0 |          0 |             0
tree.node_val_list per block *LA       |        n/a |  1.412_514 |  1.606_290 |  2.894_259 |     4.800_511
                                       |            |            |            |            |
index.bytes_read *C                    |    0.000 T |    1.949 T |    6.791 T |   15.083 T |      29.132 T
index.bytes_read per block *LA         |        n/a |    9.451 M |   20.271 M |   31.753 M |      54.879 M
index.bytes_read per sec *LA           |        n/a |  140.771 M |  144.128 M |  136.351 M |     128.626 M
index.nb_reads *C                      |    0.000 M |  538.390 M | 1363.607 M | 2355.447 M |    3604.103 M
index.nb_reads per block *LA           |        n/a |   2044.656 |   2739.774 |   3126.620 |      4459.369
index.nb_reads per sec *LA             |        n/a |     30_455 |     19_480 |     13_426 |        10_452
index.bytes_written *C                 |    0.000 T |    0.607 T |    3.261 T |    9.320 T |      20.580 T
index.bytes_written per block *LA      |        n/a |    3.796 M |   11.607 M |   23.318 M |      46.457 M
index.bytes_written per sec *LA        |        n/a |   56.537 M |   82.524 M |  100.133 M |     108.887 M
index.nb_writes *C                     |         19 |  1_589_261 |  3_289_249 |  5_038_345 |     6_933_419
index.nb_writes per block *LA          |        n/a |  4.858_136 |  5.441_777 |  5.558_851 |     5.790_718
index.nb_writes per sec *LA            |        n/a |     72.362 |     38.691 |     23.871 |        13.572
index.nb_merge *C                      |          0 |        324 |        755 |      1_279 |         1_903
                                       |            |            |            |            |
gc.minor_words allocated *C            |    0.000 T |    1.446 T |    4.694 T |   10.052 T |      18.370 T
gc.minor_words allocated per block *LA |        n/a |    6.577 M |   12.108 M |   18.868 M |      32.311 M
gc.minor_words allocated per sec *LA   |        n/a |   97.960 M |   86.089 M |   81.023 M |      75.731 M
gc.promoted_words *C                   |    0.000 G |   47.428 G |  122.658 G |  215.213 G |     329.477 G
gc.major_words allocated *C            |    0.001 G |  124.727 G |  536.367 G | 1397.732 G |    2941.268 G
gc.major_words allocated per block *LA |        n/a |    663_964 |  1_715_571 |  3_243_565 |     6_301_567
gc.major_words allocated per sec *LA   |        n/a |    9.890 M |   12.198 M |   13.928 M |      14.770 M
gc.minor_collections *C                |    0.000 M |    5.536 M |   17.966 M |   38.459 M |      70.255 M
gc.minor_collections per block *LA     |        n/a |     25.179 |     46.332 |     72.184 |       123.448
gc.minor_collections per sec *LA       |        n/a |    375.044 |    329.414 |    309.972 |       289.341
gc.major_collections *C                |          1 |      7_575 |     27_197 |     58_358 |        98_374
gc.major_collections per block *LA     |        n/a |  0.037_021 |  0.080_940 |  0.117_803 |     0.128_702
gc.major_collections per sec *LA       |        n/a |  0.551_435 |  0.575_481 |  0.505_869 |     0.301_655
gc.compactions *C                      |          0 |        164 |        386 |        662 |           987
gc.major heap bytes top *C             |    7.172 M | 1984.500 M | 4547.162 M | 4547.162 M |    4547.162 M
gc.major heap bytes *LA                |    7.172 M |  648.729 M |  892.735 M | 1059.946 M |     535.695 M
                                       |            |            |            |            |
index_data bytes *S                    |    0.000 G |    3.654 G |    8.546 G |   14.461 G |      21.520 G
index_data bytes per block *LA         |        n/a |     12_547 |     15_941 |     18_663 |        24_901
index_data bytes per sec *LA           |        n/a |    186_894 |    113_339 |     80_142 |        58_363
store_pack bytes *S                    |    0.000 G |    7.884 G |   19.594 G |   34.066 G |      51.802 G
store_pack bytes per block *LA         |        n/a |     28_265 |     38_251 |     45_745 |        62_508
store_pack bytes per sec *LA           |        n/a |    421_006 |    271_964 |    196_439 |       146_509
index_log bytes *S                     |    0.000 M |   22.500 M |   22.500 M |   22.500 M |      22.500 M
index_log_async *S                     |    0.000 M |   22.500 M |   22.500 M |   22.500 M |      22.500 M
store_dict bytes *S                    |         24 |  2_458_592 |  2_458_592 |  2_458_592 |     2_458_592
                                       |            |            |            |            |
/data/contracts/index *S               |        n/a |          4 |        256 |        256 |           256
/data/big_maps/index *S                |        n/a |          0 |         15 |         52 |           205
/data/rolls/index *S                   |        n/a |        256 |        256 |        256 |           256
/data/rolls/owner/current *S           |        n/a |        256 |        256 |        256 |           256
/data/commitments *S                   |        n/a |        256 |        256 |        256 |           256
/data/contracts/index/ed25519 *S       |        n/a |        256 |          0 |          0 |             0
/data/contracts/index/originated *S    |        n/a |        256 |          0 |          0 |             0

Types of curves:
  *C: Cumulative. No smoothing.
  *LA: Local Average. Smoothed using a weighted sum of the value in the
       block and the exponentially decayed values of the previous blocks.
       Every 16793.71 blocks, half of the past is forgotten.
  *S: Size. E.g. directory entries, file bytes. No smoothing.
```

A multi summary pretty print:
```
dune exec -- ./bench/irmin-pack/trace_stats.exe pp /data/ngoguey/irmin/_artefacts/master_none/boostrap_summary.json   /data/ngoguey/irmin/_artefacts/master_v0+v1/boostrap_summary.json
                | 0                         | 1
Hostname        | comanche                  | comanche
Word Size       | 64 bits                   | 64 bits
Start Time      | 2021/04/15 19:12:54 (GMT) | 2021/04/19 19:19:24 (GMT)
Inode Config    | mls:32 bf:32 sh:256       | mls:32 bf:32 sh:256
Store Type      | pack                      | pack
Path Conversion | none                      | v0+v1

Block played count *C                  |   | 0 (before) |    450071.2     |    900142.3     |  1343496 (end)
Blocks progress *C                     |   |     0%     |       34%       |       67%       |      100%
                                       |   |            |                 |                 |
Wall time elapsed *C                   | 0 |        0us |    6h18min      |         1d      |      2d14h
                                       | 1 |        0us |    3h11min  50% |   10h36min  43% |   23h49min  38%
CPU time elapsed *C                    | 0 |        0us |    6h14min      |         1d      |      2d13h
                                       | 1 |        0us |    3h11min  51% |   10h26min  43% |   23h29min  38%
                                       |   |            |                 |                 |
Ops count *C                           | 0 |    0.000 M |  250.775 M      |  592.948 M      | 1089.854 M
                                       | 1 |    0.000 M |  250.775 M 100% |  592.948 M 100% | 1089.854 M 100%
Read only ops per sec *LA              | 0 |        n/a |      5_638      |      3_184      |      2_492
                                       | 1 |        n/a |     10_843 192% |      7_966 250% |      6_870 276%
Read write ops per sec *LA             | 0 |        n/a |   2119.146      |   1105.795      |    784.451
                                       | 1 |        n/a |   4075.949 192% |   2766.265 250% |   2162.382 276%
                                       |   |            |                 |                 |
Commit every sec *LA                   | 0 |        n/a |     12.208      |      5.000      |      2.344
                                       | 1 |        n/a |     23.480 192% |     12.508 250% |      6.461 276%
Add every sec *LA                      | 0 |        n/a |   2102.175      |   1097.681      |    773.094
                                       | 1 |        n/a |   4043.305 192% |   2745.969 250% |   2131.077 276%
Remove every sec *LA                   | 0 |        n/a |     16.924      |      8.093      |     11.123
                                       | 1 |        n/a |     32.551 192% |     20.247 250% |     30.661 276%
Find every sec *LA                     | 0 |        n/a |   4178.387      |   2274.828      |   1663.274
                                       | 1 |        n/a |   8036.676 192% |   5690.728 250% |   4584.906 276%
Mem every sec *LA                      | 0 |        n/a |   1056.270      |    744.592      |    751.784
                                       | 1 |        n/a |   2031.620 192% |   1862.677 250% |   2072.335 276%
Mem_tree every sec *LA                 | 0 |        n/a |    402.851      |    165.002      |     77.346
                                       | 1 |        n/a |    774.841 192% |    412.771 250% |    213.209 276%
Copy every sec *LA                     | 0 |        n/a |  0.047_900      |  0.019_613      |  0.233_792
                                       | 1 |        n/a |  0.092_130 192% |  0.049_065 250% |  0.644_460 276%
                                       |   |            |                 |                 |
Add count per block *LA                | 0 |        n/a |    172.202      |    219.537      |    329.843
                                       | 1 |        n/a |    172.202 100% |    219.537 100% |    329.843 100%
Remove count per block *LA             | 0 |        n/a |  1.386_346      |  1.618_699      |  4.745_588
                                       | 1 |        n/a |  1.386_346 100% |  1.618_699 100% |  4.745_588 100%
Find count per block *LA               | 0 |        n/a |    342.277      |    454.967      |    709.640
                                       | 1 |        n/a |    342.277 100% |    454.967 100% |    709.640 100%
Mem count per block *LA                | 0 |        n/a |     86.525      |    148.919      |    320.751
                                       | 1 |        n/a |     86.525 100% |    148.919 100% |    320.751 100%
Mem_tree count per block *LA           | 0 |        n/a |     33.000      |     33.001      |     33.000
                                       | 1 |        n/a |     33.000 100% |     33.001 100% |     33.000 100%
Copy count per block *LA               | 0 |        n/a |  0.003_924      |  0.003_923      |  0.099_748
                                       | 1 |        n/a |  0.003_924 100% |  0.003_923 100% |  0.099_748 100%
                                       |   |            |                 |                 |
Block duration *LA                     | 0 |   0.000  s |   0.082  s      |   0.200  s      |   0.427  s
                                       | 1 |   0.000  s |   0.043  s  52% |   0.080  s  40% |   0.155  s  36%
Buildup seen duration *LA              | 0 |   0.000 ms |   5.776 ms      |   8.010 ms      |  12.523 ms
                                       | 1 |   0.000 ms |   3.905 ms  68% |   4.992 ms  62% |   7.768 ms  62%
Buildup unseen duration *LA            | 0 |        n/a |   2.015 ms      |   2.659 ms      |   4.478 ms
                                       | 1 |        n/a |   3.152 ms 156% |   3.492 ms 131% |   5.222 ms 117%
Commit duration *LA                    | 0 |        n/a |   0.074  s      |   0.189  s      |   0.410  s
                                       | 1 |        n/a |   0.036  s  48% |   0.071  s  38% |   0.142  s  35%
                                       |   |            |                 |                 |
Add duration *LA                       | 0 |        n/a |  11.870 µs      |  10.397 µs      |   9.556 µs
                                       | 1 |        n/a |   7.403 µs  62% |   6.228 µs  60% |   5.462 µs  57%
Remove duration *LA                    | 0 |        n/a |  11.996 µs      |  30.224 µs      |  32.297 µs
                                       | 1 |        n/a |   6.425 µs  54% |  14.448 µs  48% |  15.280 µs  47%
Find duration *LA                      | 0 |        n/a |   9.696 µs      |   9.233 µs      |   9.272 µs
                                       | 1 |        n/a |   6.659 µs  69% |   5.794 µs  63% |   5.983 µs  65%
Mem duration *LA                       | 0 |        n/a |   3.597 µs      |   9.274 µs      |   7.912 µs
                                       | 1 |        n/a |   2.693 µs  75% |   6.023 µs  65% |   4.931 µs  62%
Mem_tree duration *LA                  | 0 |        n/a |   2.439 µs      |   2.787 µs      |   2.829 µs
                                       | 1 |        n/a |   3.150 µs 129% |   1.886 µs  68% |   1.781 µs  63%
Copy duration *LA                      | 0 |        n/a |  18.139 µs      |  38.074 µs      |  13.476 µs
                                       | 1 |        n/a |  16.061 µs  89% |  14.802 µs  39% |  11.111 µs  82%
Checkout duration *LA                  | 0 |        n/a |   4.571 µs      |   4.527 µs      |   5.624 µs
                                       | 1 |        n/a |   5.143 µs 113% |   6.044 µs 134% |   6.500 µs 116%
                                       |   |            |                 |                 |
pack.finds per block *LA               | 0 |        n/a |   1710.372      |   2225.523      |   2859.420
                                       | 1 |        n/a |    630.879  37% |    739.917  33% |    955.487  33%
pack.cache_misses per block *LA        | 0 |        n/a |     88.360      |    121.048      |    206.057
                                       | 1 |        n/a |     33.407  38% |     53.979  45% |     93.906  46%
pack.appended_hashes per block *LA     | 0 |        n/a |  0.064_756      |  0.133_749      |  0.073_375
                                       | 1 |        n/a |  0.064_756 100% |  0.133_749 100% |  0.073_375 100%
pack.appended_offsets per block *LA    | 0 |        n/a |   1755.590      |   2463.144      |   3576.880
                                       | 1 |        n/a |   1941.100 111% |   2857.382 116% |   4194.988 117%
                                       |   |            |                 |                 |
tree.contents_hash per block *LA       | 0 |        n/a |    132.200      |    145.706      |    176.406
                                       | 1 |        n/a |    132.200 100% |    145.706 100% |    176.406 100%
tree.contents_find per block *LA       | 0 |        n/a |    195.419      |    248.149      |    348.266
                                       | 1 |        n/a |    195.419 100% |    248.161 100% |    348.359 100%
tree.contents_add per block *LA        | 0 |        n/a |     73.823      |     86.314      |    107.554
                                       | 1 |        n/a |     73.896 100% |     86.425 100% |    107.750 100%
tree.node_hash per block *LA           | 0 |        n/a |    222.388      |    296.157      |    412.139
                                       | 1 |        n/a |     82.650  37% |     86.508  29% |    115.339  28%
tree.node_mem per block *LA            | 0 |        n/a |    221.439      |    294.614      |    407.885
                                       | 1 |        n/a |     82.518  37% |     86.318  29% |    114.910  28%
tree.node_add per block *LA            | 0 |        n/a |    218.287      |    292.239      |    404.644
                                       | 1 |        n/a |     79.566  36% |     84.330  29% |    112.608  28%
tree.node_find per block *LA           | 0 |        n/a |    705.821      |    803.104      |    965.929
                                       | 1 |        n/a |    247.814  35% |    263.319  33% |    314.188  33%
tree.node_val_v per block *LA          | 0 |        n/a |  7.065e-13      |  6.047e-21      |  6.829e-29
                                       | 1 |        n/a |  7.065e-13 100% |  6.047e-21 100% |  6.829e-29 100%
tree.node_val_find per block *LA       | 0 |        n/a |          0      |          0      |          0
                                       | 1 |        n/a |          0      |          0      |          0
tree.node_val_list per block *LA       | 0 |        n/a |  1.690_392      |  2.235_998      |  4.800_511
                                       | 1 |        n/a |  1.157_440  68% |  1.190_748  53% |  1.514_817  32%
                                       |   |            |                 |                 |
index.bytes_read *C                    | 0 |    0.000 T |    3.149 T      |   11.840 T      |   29.132 T
                                       | 1 |    0.000 T |    2.134 T  68% |    6.756 T  57% |   14.625 T  50%
index.bytes_read per block *LA         | 0 |        n/a |   10.182 M      |   26.096 M      |   54.879 M
                                       | 1 |        n/a |    6.652 M  65% |   13.610 M  52% |   25.444 M  46%
index.bytes_read per sec *LA           | 0 |        n/a |  124.298 M      |  130.477 M      |  128.626 M
                                       | 1 |        n/a |  156.181 M 126% |  170.238 M 130% |  164.390 M 128%
index.nb_reads *C                      | 0 |    0.000 M |  788.413 M      | 2022.180 M      | 3604.103 M
                                       | 1 |    0.000 M |  636.667 M  81% | 1661.028 M  82% | 3088.435 M  86%
index.nb_reads per block *LA           | 0 |        n/a |   2216.740      |   2947.196      |   4459.369
                                       | 1 |        n/a |   1857.579  84% |   2523.268  86% |   4051.360  91%
index.nb_reads per sec *LA             | 0 |        n/a |     27_061      |     14_736      |     10_452
                                       | 1 |        n/a |     43_616 161% |     31_561 214% |     26_175 250%
index.bytes_written *C                 | 0 |    0.000 T |    1.185 T      |    6.940 T      |   20.580 T
                                       | 1 |    0.000 T |    0.399 T  34% |    2.260 T  33% |    6.629 T  32%
index.bytes_written per block *LA      | 0 |        n/a |    5.607 M      |   19.195 M      |   46.457 M
                                       | 1 |        n/a |    2.001 M  36% |    6.265 M  33% |   15.882 M  34%
index.bytes_written per sec *LA        | 0 |        n/a |   68.451 M      |   95.977 M      |  108.887 M
                                       | 1 |        n/a |   46.979 M  69% |   78.360 M  82% |  102.612 M  94%
index.nb_writes *C                     | 0 |         19 |  2_140_791      |  4_448_755      |  6_933_419
                                       | 1 |         19 |  1_990_265  93% |  4_057_497  91% |  6_167_447  89%
index.nb_writes per block *LA          | 0 |        n/a |  4.728_969      |  5.271_985      |  5.790_718
                                       | 1 |        n/a |  4.446_586  94% |  4.736_967  90% |  5.010_228  87%
index.nb_writes per sec *LA            | 0 |        n/a |     57.729      |     26.360      |     13.572
                                       | 1 |        n/a |    104.406 181% |     59.250 225% |     32.371 239%
index.nb_merge *C                      | 0 |          0 |        454      |      1_103      |      1_903
                                       | 1 |          0 |        260  57% |        626  57% |      1_076  57%
                                       |   |            |                 |                 |
gc.minor_words allocated *C            | 0 |    0.000 T |    2.300 T      |    8.088 T      |   18.370 T
                                       | 1 |    0.000 T |    1.071 T  47% |    3.153 T  39% |    6.801 T  37%
gc.minor_words allocated per block *LA | 0 |        n/a |    7.770 M      |   16.512 M      |   32.311 M
                                       | 1 |        n/a |    3.355 M  43% |    5.890 M  36% |   11.875 M  37%
gc.minor_words allocated per sec *LA   | 0 |        n/a |   94.856 M      |   82.558 M      |   75.731 M
                                       | 1 |        n/a |   78.784 M  83% |   73.669 M  89% |   76.725 M 101%
gc.promoted_words *C                   | 0 |    0.000 G |   69.636 G      |  183.911 G      |  329.477 G
                                       | 1 |    0.000 G |   36.201 G  52% |   86.016 G  47% |  151.900 G  46%
gc.major_words allocated *C            | 0 |    0.001 G |  220.272 G      | 1065.090 G      | 2941.268 G
                                       | 1 |    0.001 G |   87.045 G  40% |  373.020 G  35% |  993.472 G  34%
gc.major_words allocated per block *LA | 0 |        n/a |    910_013      |  2_710_317      |  6_301_567
                                       | 1 |        n/a |    351_540  39% |    914_365  34% |  2_198_077  35%
gc.major_words allocated per sec *LA   | 0 |        n/a |   11.109 M      |   13.552 M      |   14.770 M
                                       | 1 |        n/a |    8.254 M  74% |   11.437 M  84% |   14.202 M  96%
gc.minor_collections *C                | 0 |    0.000 M |    8.804 M      |   30.945 M      |   70.255 M
                                       | 1 |    0.000 M |    4.099 M  47% |   12.074 M  39% |   26.031 M  37%
gc.minor_collections per block *LA     | 0 |        n/a |     29.732      |     63.156      |    123.448
                                       | 1 |        n/a |     12.850  43% |     22.548  36% |     45.434  37%
gc.minor_collections per sec *LA       | 0 |        n/a |    362.954      |    315.777      |    289.341
                                       | 1 |        n/a |    301.712  83% |    282.030  89% |    293.546 101%
gc.major_collections *C                | 0 |          1 |     12_316      |     46_096      |     98_374
                                       | 1 |          1 |      5_031  41% |     18_238  40% |     39_634  40%
gc.major_collections per block *LA     | 0 |        n/a |  0.043_392      |  0.099_340      |  0.128_702
                                       | 1 |        n/a |  0.018_733  43% |  0.040_777  41% |  0.065_500  51%
gc.major_collections per sec *LA       | 0 |        n/a |  0.529_714      |  0.496_700      |  0.301_655
                                       | 1 |        n/a |  0.439_847  83% |  0.510_036 103% |  0.423_190 140%
gc.compactions *C                      | 0 |          0 |        232      |        571      |        987
                                       | 1 |          0 |        138  59% |        322  56% |        547  55%
gc.major heap bytes top *C             | 0 |    7.172 M | 2493.854 M      | 4547.162 M      | 4547.162 M
                                       | 1 |    7.172 M | 2054.713 M  82% | 2622.525 M  58% | 3530.043 M  78%
gc.major heap bytes *LA                | 0 |    7.172 M |  849.887 M      | 1066.414 M      |  535.695 M
                                       | 1 |    7.172 M |  671.678 M  79% |  825.680 M  77% | 1141.486 M 213%
                                       |   |            |                 |                 |
index_data bytes *S                    | 0 |    0.000 G |    5.126 G      |   12.480 G      |   21.520 G
                                       | 1 |    0.000 G |    2.934 G  57% |    7.061 G  57% |   12.143 G  56%
index_data bytes per block *LA         | 0 |        n/a |     13_007      |     17_869      |     24_901
                                       | 1 |        n/a |      7_995  61% |     10_214  57% |     15_020  60%
index_data bytes per sec *LA           | 0 |        n/a |    158_780      |     89_347      |     58_363
                                       | 1 |        n/a |    187_729 118% |    127_758 143% |     97_040 166%
store_pack bytes *S                    | 0 |    0.000 G |   11.320 G      |   29.187 G      |   51.802 G
                                       | 1 |    0.000 G |   11.800 G 104% |   34.633 G 119% |   71.232 G 138%
store_pack bytes per block *LA         | 0 |        n/a |     30_421      |     43_427      |     62_508
                                       | 1 |        n/a |     34_217 112% |     61_092 141% |     94_505 151%
store_pack bytes per sec *LA           | 0 |        n/a |    371_363      |    217_132      |    146_509
                                       | 1 |        n/a |    803_414 216% |    764_144 352% |    610_589 417%
index_log bytes *S                     | 0 |    0.000 M |   22.500 M      |   22.500 M      |   22.500 M
                                       | 1 |    0.000 M |   22.500 M 100% |   22.500 M 100% |   22.500 M 100%
index_log_async *S                     | 0 |    0.000 M |   22.500 M      |   22.500 M      |   22.500 M
                                       | 1 |    0.000 M |   22.500 M 100% |   22.500 M 100% |   22.500 M 100%
store_dict bytes *S                    | 0 |         24 |  2_458_592      |  2_458_592      |  2_458_592
                                       | 1 |         24 |  3_074_852 125% |  3_074_852 125% |  3_074_852 125%
                                       |   |            |                 |                 |
/data/contracts/index *S               | 0 |        n/a |          4      |        256      |        256
                                       | 1 |        n/a |          4 100% |    464_138 2e3x |  1_016_194 4e3x
/data/big_maps/index *S                | 0 |        n/a |          0      |         31      |        205
                                       | 1 |        n/a |          0      |         32 103% |        417 203%
/data/rolls/index *S                   | 0 |        n/a |        256      |        256      |        256
                                       | 1 |        n/a |     55_938 219x |     81_987 320x |     85_410 334x
/data/rolls/owner/current *S           | 0 |        n/a |        256      |        256      |        256
                                       | 1 |        n/a |     56_401 220x |     81_118 317x |     84_614 331x
/data/commitments *S                   | 0 |        n/a |        256      |        256      |        256
                                       | 1 |        n/a |     13_445  53x |      8_287  32x |      5_935  23x
/data/contracts/index/ed25519 *S       | 0 |        n/a |        256      |          0      |          0
                                       | 1 |        n/a |    247_726 968x |          0      |          0
/data/contracts/index/originated *S    | 0 |        n/a |        256      |          0      |          0
                                       | 1 |        n/a |     18_856  74x |          0      |          0

Types of curves:
  *C: Cumulative. No smoothing.
  *LA: Local Average. Smoothed using a weighted sum of the value in the
       block and the exponentially decayed values of the previous blocks.
       Every 16793.71 blocks, half of the past is forgotten.
  *S: Size. E.g. directory entries, file bytes. No smoothing.
```

### CODE REVIEW

**Comment by samoht** [comment]
That looks great! Would be super useful to start the summary with these metrics:

- NTS (number of transactions per seconds)
- IOPS (total number of IO operations (read + writes) per seconds, and only reads and only writes)
- throughput (MB/s for read+writes and reads and writes)
- Memory usage (with rusage?)

I believe these are the main metrics we are interested in right now.

**Comment by Ngoguey42** [comment]
@samoht thanks for your inputs. I will reorder and add the missing stats in another PR.

---

> NTS (number of transactions per seconds)

This will require a new replay trace. This won't be available soon.

---

> Memory usage (with rusage?)

I will add that to the stat trace. Note that the size of the major heap is already present:
```
gc.major heap bytes top *C             |    7.172 M | 1984.500 M | 4547.162 M | 4547.162 M |    4547.162 M
gc.major heap bytes *LA                |    7.172 M |  648.729 M |  892.735 M | 1059.946 M |     535.695 M
```

---

> IOPS (total number of IO operations (read + writes) per seconds, and only reads and only writes)
> throughput (MB/s for read+writes and reads and writes)

Half of those stats are already available in there.

Two thoughts about those `per second` stats:

I will remove the current `per sec` stats as shown so far and print their average over the full run instead. The reason is because I noticed that those numbers are **extremely** noisy:
- they are computed `per block` and then converted `per second`;
- the wall clock is itself a noisy stats on our non-stabilized systems.

When possible, we should favor the `per block` stats. For example:
- Lowering by 20% the number of reads `per second` could be a good sign, if the number of blocks per seconds improved, or a bad sign otherwise.
- Lowering by 20% the number of reads `per block` is always good thing, not subject to intepretation :)

---

**Comment by samoht** [comment]
> This will require a new replay trace. This won't be available soon.

So initially let's compute that metrics from IOPS (with something like 1 transaction = 3 operations or something like this). We can use the real data when it's available later on.

For the memory usage metrics: we need min / mean /max usage not cumulative stats. Is it possible to get these?

For the per x seconds metrics: it's actually what we need to improve, so I really want to see those appear. The blocks can contain various operations (sometime not a lot, sometimes a lot) so that metrics is not the most important one (it's nice to have it somewhere below).

File: bench/irmin-pack/trace_stat_summary_utils.ml [hunk 1]
```diff
@@ -17,6 +17,100 @@
 type histo = (float * int) list [@@deriving repr]
 type curve = float list [@@deriving repr]
 
+let snap_to_integer ~significant_digits v =
```
Comment (icristescu) [inline bench/irmin-pack/trace_stat_summary_utils.ml:20]: It's not clear to me what the `~significant_digits` are, can you maybe add a comment for this function?

File: bench/irmin-pack/trace_stat_summary_utils.mli [hunk 2]
```diff
@@ -6,6 +6,32 @@
 type histo = (float * int) list [@@deriving repr]
 type curve = float list [@@deriving repr]
 
+val create_pp_real :
+  ?significant_digits:int -> float list -> Format.formatter -> float -> unit
+(** [create_pp_real examples] is [pp_real], a float pretty-printer that adapts
+    to the [examples] shown to it.
+
+    It is highly recommended, but not mandatory, for all the numbers passed to
+    [pp_real] to be included [examples].
```
Comment (icristescu) [inline bench/irmin-pack/trace_stat_summary_utils.mli:15]:
```suggestion
    [pp_real] to be included in [examples].
```

File: bench/irmin-pack/trace_stat_summary_utils.mli [hunk 3]
```diff
@@ -6,6 +6,32 @@
...
+
+    [significant_digits] is used to snap certain numbers to the nearest integer. *)
```
Comment (icristescu) [inline bench/irmin-pack/trace_stat_summary_utils.mli:17]: `round` instead of `snap`?

**Review by icristescu — APPROVED** [review]
just nitpicks, it looks great!

File: bench/irmin-pack/trace_stat_summary_utils.mli (hunk 3 above)
Comment (Ngoguey42) [reply bench/irmin-pack/trace_stat_summary_utils.mli:17]: `round` would suggest that all the output numbers are integers. `snap` is a common naming scheme for such behaviours.

File: bench/irmin-pack/trace_stat_summary_utils.ml (hunk 1 above)
Comment (Ngoguey42) [reply bench/irmin-pack/trace_stat_summary_utils.ml:20]: I will, in the next PR.

Comment (Ngoguey42) [reply bench/irmin-pack/trace_stat_summary_utils.ml:20]: Documenting it made me find bugs.

See https://github.com/mirage/irmin/pull/1416/commits/dcf08cdb433cd95e945ead6c872ab050c8d18685

**Comment by Ngoguey42** [comment]
I will squash and merge when the CI approves

File: bench/irmin-pack/trace_stat_summary_utils.ml (hunk 1 above)
Comment (Ngoguey42) [reply bench/irmin-pack/trace_stat_summary_utils.ml:20]: And https://github.com/mirage/irmin/pull/1416/commits/a5416193c88c248eb3f5b55220258710fc809bdc

**Pull request merged**